File: PyMungBean/src/moments/ellipse.py
'''
Created on May 4, 2011

@author: anol
'''
import cv

WIDTH = 101
HEIGTH = 101

WHITE = cv.Scalar(0xff, 0xff, 0xff, 0)
BLACK = cv.Scalar(0, 0, 0, 0)

def xfrange(start, stop, step):
    while start < stop:
        yield start
        start += step

def Ellipse(img, size = (80,50)):
    ellip = cv.CloneImage(img)

    ellip_center = (cv.Round(WIDTH * 0.5) + 1, cv.Round(HEIGTH * 0.5) + 1)
    ellip_size = size
    cv.Zero(ellip)
    box = (ellip_center, ellip_size, 0.0)
    cv.EllipseBox(ellip, box, WHITE, 1, 8, 0)
    storage = cv.CreateMemStorage()
    seq = cv.FindContours(ellip, storage, cv.CV_RETR_LIST, cv.CV_CHAIN_APPROX_SIMPLE)
    cv.DrawContours(img, seq, WHITE, BLACK, -1, cv.CV_FILLED, 8)
    return img

if __name__ == '__main__':
    size = (WIDTH, HEIGTH)
    ellip = cv.CreateImage(size, cv.IPL_DEPTH_8U, 1)

    ellip_center = (cv.Round(WIDTH * 0.5) + 1, cv.Round(HEIGTH * 0.5) + 1)
    ellip_size = (80, 50)

    for deg in xfrange(0.0, 360.0, 0.5):
        cv.Zero(ellip)
        box = (ellip_center, ellip_size, deg)
        cv.EllipseBox(ellip, box, WHITE, 1, 8, 0)
        img = cv.CloneImage(ellip)
        cv.Zero(img)

        storage = cv.CreateMemStorage()
        seq = cv.FindContours(ellip, storage, cv.CV_RETR_LIST, cv.CV_CHAIN_APPROX_SIMPLE)
        cv.DrawContours(img, seq, WHITE, BLACK, -1, cv.CV_FILLED, 8)

        moment = cv.Moments(img, 1)
        hu = cv.GetHuMoments(moment)

        print "%3.2f\t%+8.3e\t%+8.3e\t%+8.3e\t%+8.3e\t%+8.3e\t%+8.3e\t%+8.3e\n" % (deg, hu[0], hu[1], hu[2], hu[3], hu[4], hu[5], hu[6])
#        cv.ShowImage('Rectangular', rect)
        cv.ShowImage('Box', img)
        if cv.WaitKey(10) == 27:
            break
    cv.DestroyWindow('Box')

File: PyMungBean/src/moments/Ellipse.c
/*
 ============================================================================
 Name        : Moment.c
 Author      : Anol
 Version     :
 Copyright   : Your copyright notice
 Description : Hello World in C, Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>

#include <opencv/cxcore.h>
#include <opencv/cv.h>
#include <opencv/highgui.h>
#include <opencv/cxtypes.h>

#define WIDTH 101
#define HEIGTH 101

#define WHITE cvScalar(0xff, 0xff, 0xff, 0)
#define BLACK cvScalar(0, 0, 0, 0)

int main(void) {

	//	/*	Rectangular	*/
	CvSize size = cvSize(WIDTH, HEIGTH);
	//	IplImage * rect = cvCreateImage(size, IPL_DEPTH_8U, 1);
	//	CvRect rect_size = cvRect(50, 50, 400, 400);
	//	CvScalar rect_clr = cvScalar(0xff, 0xff, 0, 0);
	//		cvRectangleR(rect, rect_size, rect_clr, 1, 8, 0);

	/*	Ellipse	*/
	IplImage * ellip = cvCreateImage(size, IPL_DEPTH_8U, 1);
	CvPoint2D32f ellip_center = cvPoint2D32f((WIDTH * 0.5) + 1,
			(HEIGTH * 0.5) + 1);
	CvSize2D32f ellip_size = cvSize2D32f(80, 50);

	/*	Triangle	*/

	//	IplImage *img = cvCloneImage(ellip);
	//	cvZero(img);

	for (float deg = 0; deg <= 180; deg += 0.1) {
		cvZero(ellip);
		CvBox2D ellip_box = { ellip_center, ellip_size, (float) deg };
		cvEllipseBox(ellip, ellip_box, WHITE, 1, 8, 0);
		IplImage *img = cvCloneImage(ellip);
		cvZero(img);

		CvMemStorage* storage = cvCreateMemStorage(0);
		CvSeq* contour = NULL;

		cvFindContours(ellip, storage, &contour, sizeof(CvContour),
				CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE, cvPoint(0, 0));

		//		for (CvSeq* c = contour; c != NULL; c = c->h_next) {

		cvDrawContours(img, contour, WHITE, BLACK, -1, CV_FILLED, 8,
				cvPoint(0, 0));
		//		}

		CvMoments moment;
		CvHuMoments hu;
		cvMoments(img, &moment, 1);
		cvGetHuMoments(&moment, &hu);

		//		cvShowImage("Rectangular", rect);
		printf(
				"%3.2f : %+8.3E %+8.3E %+8.3E %+8.3E %+8.3E %+8.3E %+8.3E\n",
				deg, hu.hu1, hu.hu2, hu.hu3, hu.hu4, hu.hu5, hu.hu6, hu.hu7);

		cvShowImage("Ellipse", ellip);
		cvShowImage("Ellipse Contour", img);

		int ret = cvWaitKey(10);
		if (ret == 27)
			break;
		//		if (cvWaitKey(0) == 27)
		//			;
		//		break;
		//		cvWaitKey(0);
	}

	cvDestroyAllWindows();
	puts("!!!Application Closed!!!"); /* prints !!!Hello World!!! */
	return EXIT_SUCCESS;
}
